### PR TITLE
Documentation update idea

### PR DESCRIPTION
**Pull request by marcinczeczko**
## WORK IN PROGRESS
---

I hereby agree to the terms of the Knot.x Contributor License Agreement.

### CODE REVIEW

File: documentation/src/main/wiki/CoreArchitecture.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+[WORKING IN PROGRESS]
+#Knot.x Core Architecture
+
+Diagram below depicts Knot.x core modules. Knot.x by default comes with Core and Core Adapters modules.
+Every module can be disabled / replaced with a simple JSON configuration change.
+
+[[assets/knotx-architecture-full.png|alt=Knot.x Core Architecture]]
```
Comment (wiiitek) [inline documentation/src/main/wiki/CoreArchitecture.md]: We are adding some png images. They are probably created with some diagrams editor and exported to png format.

Could we attach original format for the diagrams editor in order to update those images easily?

File: documentation/src/main/wiki/CommunicationFlow.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
+[TO REVIEW]
+
+# Communication Flow
+
+A diagram below depicts a request flow inside Knot.x:
+
+[[assets/knotx-modules-request-flow.png|alt=Knot.x Request Flow]]
```
Comment (Skejven) [inline documentation/src/main/wiki/CommunicationFlow.md]: I had exactly the same idea. Simpler flow added.

File: documentation/src/main/wiki/ActionAdapter.md [hunk 3]
```diff
@@ -0,0 +1,16 @@
+Action Adapter is Component of a system, that mediate communication between Knot.x (ActionKnot) and external 
+Services that are responsible for handling form submitting.
```
Comment (marcinczeczko) [inline documentation/src/main/wiki/ActionAdapter.md]: ... form submission

File: documentation/src/main/wiki/ActionAdapter.md [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+Action Adapter is a project custom part of a system, however Knot.x comes with default `HttpActionAdapter`.
```
Comment (marcinczeczko) [inline documentation/src/main/wiki/ActionAdapter.md]: Action Adapter is an extension of Knot.x. By delivering custom implementations of Action Adapters, that meet your project requirements, it allows you to adopt request from Knot.x into request understandable by endpoint service, and adopts responses from that service into unified message understandable by Knot.x.
For you convenience, Knot.x is shipped with a default Action Adapter called `HttpActionAdapter`.

File: documentation/src/main/wiki/ActionAdapter.md [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+Action Adapter entry are:
```
Comment (marcinczeczko) [inline documentation/src/main/wiki/ActionAdapter.md]: Action Adapter accepts message with the following data:
....

File: documentation/src/main/wiki/ActionAdapter.md [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+- `ClientRequest` - object with all data of an original request,
+- `params` - all additional params defined in configuration.
+
+Action Adapter result contract is Json Object with fields:
```
Comment (marcinczeczko) [inline documentation/src/main/wiki/ActionAdapter.md]: Result generated by Action Adapter must be a JsonObject with the fields as below:

File: documentation/src/main/wiki/ActionKnot.md [hunk 7]
```diff
@@ -0,0 +1,35 @@
+ActionKnot is Knot.x system component that is responsible for handling forms submitting.
```
Comment (marcinczeczko) [inline documentation/src/main/wiki/ActionKnot.md]: ..forms submission.

File: documentation/src/main/wiki/ActionKnot.md [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    <input type="email" name="email" value="{{#if action._result.validationError}} {{action._result.form.email}} {{/if}}" />
+    <input type="submit" value="Submit"/>
+  </form>
+  <div>
+    <strong>Pro tip: All emails that starts with <kbd>john.doe</kbd> will be accepted.</strong>
+  </div>
+</script>
+```
+
+In order to enable ActionKnot interpretation of a fragment, `data-api-type="form-{NAME}"` must be defined in `<script>` tag,
```
Comment (marcinczeczko) [inline documentation/src/main/wiki/ActionKnot.md]: ActionKnot processes only those ''<script>' tags that have `data-api-type=form-{name}'

File: documentation/src/main/wiki/ActionKnot.md [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    <strong>Pro tip: All emails that starts with <kbd>john.doe</kbd> will be accepted.</strong>
+  </div>
+</script>
+```
+
+In order to enable ActionKnot interpretation of a fragment, `data-api-type="form-{NAME}"` must be defined in `<script>` tag, 
+where `{NAME}` is a unique name of a form (assuming there may be more than one form on a single page it is used to distinguish requested snippet).
+In the example above, `{NAME}` is `1`.
+
+Following data attributes are available in `<form>` tag with described purpose:
```
Comment (marcinczeczko) [inline documentation/src/main/wiki/ActionKnot.md]: Following data attributes are available in the <form> tag:

File: documentation/src/main/wiki/ActionKnot.md [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+- `data-knotx-action` - this is a name of an [[Action Adapter|ActionAdapter]] that will be used to handle submitted data. 
+It is similar concept as `data-service-{NAME}` in [[View Knot|ViewKnot]]. In the example above, 
+Action Handler registered under name `step1` will handle this form data submit.
+- `data-knotx-on-{SIGNAL}` - name of a [Signal](#Sginal) that should be applied.
+
+#### Signal
+Signal is basically a decision about further request processing. Value of the signal can be either:
+
+- `url` of a page that user should be redirected to after processing form submit,
+- special value `_self` that indicates further processing of this request by consecutive [[Knots|Knot]].
```
Comment (marcinczeczko) [inline documentation/src/main/wiki/ActionKnot.md]: _self - that indicates that there will not be redirect, instead page itself will be processed (generated view for instance). In other words, the page processing will be delegated to next Knot in the graph

**Comment by Skejven** [comment]
Still ToDo:
1) update main `README.md` file.
2) update `Home` with short description and the most important values that are behind Knot.x.
3) `Getting Started` should be shorter and more friendly. Move modules description to another page.
4) Move `SnippetProcessing` content to other pages (e.g. to `ServiceAdapters` description and examples).
5) Update `Server` description.
6) Provide `Splitter` description.
7) Provide `Knot` description, provide a diagram that show how do `Knots` work.
8) Update `ActionKnot` - provide a diagram and example.
9) Update `ViewKnot` - provide a diagram and example.
10) Create `AuthorizationKnot` - provide a diagram and example.
11) Provide `Adapter` description.
12) Provide `ServiceAdapter` description with example (`HttpServiceAdapter`) and HOWTO implement custom Service Adapter.
13) Provide `ActionAdapter` description with example (`HttpActionAdapter`) and HOWTO implement custom Action Adapter.
14) Update `KnotxDeployment`, provide more info about configuring and logging.

**Pull request merged**